bee-ledger/src/workers/pruning/batch.rs

```diff
@@ -277,115 +292,223 @@ pub fn batch_prunable_unconfirmed_data<S: StorageBackend>(
     }
 
-    Ok(metrics)
+    Ok((byte_length, metrics))
 }
 
-pub fn prune_milestone_data<S: StorageBackend>(
+pub(crate) fn prune_milestone_data<S: StorageBackend>(
     storage: &S,
     batch: &mut S::Batch,
     prune_index: MilestoneIndex,
     should_prune_receipts: bool,
-) -> Result<MilestoneDataPruningMetrics, Error> {
+    by_size: bool,
+) -> Result<(ByteLength, MilestoneDataPruningMetrics), PruningError> {
+    let mut byte_length = 0usize;
     let mut metrics = MilestoneDataPruningMetrics::default();
 
-    prune_milestone(storage, batch, prune_index)?;
-
-    prune_output_diff(storage, batch, prune_index)?;
+    byte_length += prune_milestone(storage, batch, prune_index, by_size)?;
+    byte_length += prune_output_diff(storage, batch, prune_index, by_size)?;
 
     if should_prune_receipts {
-        metrics.receipts = prune_receipts(storage, batch, prune_index)?;
+        let (num_receipts, num_bytes) = prune_receipts(storage, batch, prune_index, by_size)?;
+
+        metrics.receipts = num_receipts;
+        byte_length += num_bytes;
     }
 
-    Ok(metrics)
+    Ok((byte_length, metrics))
 }
 
 fn prune_message_and_metadata<S: StorageBackend>(
     storage: &S,
     batch: &mut S::Batch,
     message_id: &MessageId,
-) -> Result<(), Error> {
-    Batch::<MessageId, Message>::batch_delete(storage, batch, message_id).map_err(|e| Error::Storage(Box::new(e)))?;
+    by_size: bool,
+) -> Result<ByteLength, PruningError> {
+    let mut byte_length = 0usize;
+
+    if by_size {
+        // Panic: we know the message is in the database, because the caller already made sure of that.
+        let msg = Fetch::<MessageId, Message>::fetch(storage, message_id)
+            .map_err(|e| PruningError::Storage(Box::new(e)))?
+            .unwrap();
+        byte_length += msg.packed_len();
+
+        // Panic: we know the message metadata is in the database for the same reason as above.
+        let md = Fetch::<MessageId, MessageMetadata>::fetch(storage, message_id)
+            .map_err(|e| PruningError::Storage(Box::new(e)))?
+            .unwrap();
+        byte_length += md.packed_len();
+    }
+
+    Batch::<MessageId, Message>::batch_delete(storage, batch, message_id)
+        .map_err(|e| PruningError::Storage(Box::new(e)))?;
+
     Batch::<MessageId, MessageMetadata>::batch_delete(storage, batch, message_id)
-        .map_err(|e| Error::Storage(Box::new(e)))?;
+        .map_err(|e| PruningError::Storage(Box::new(e)))?;
 
-    Ok(())
+    Ok(byte_length)
 }
 
 fn prune_edge<S: StorageBackend>(
     storage: &S,
     batch: &mut S::Batch,
     edge: &(MessageId, MessageId),
-) -> Result<(), Error> {
-    Batch::<(MessageId, MessageId), ()>::batch_delete(storage, batch, edge).map_err(|e| Error::Storage(Box::new(e)))?;
+    by_size: bool,
+) -> Result<ByteLength, PruningError> {
+    let mut byte_length = 0usize;
+
+    if by_size {
+        byte_length += edge.0.packed_len() + edge.1.packed_len();
```

Comment:
iirc tuples implement `Packable` so you can do `edge.packed_len()` but I might be wrong

Reply:
changed